src/v/s3/signature.cc

```diff
@@ -0,0 +1,392 @@
+#include "s3/signature.h"
+
+#include "s3/error.h"
+#include "s3/logger.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <boost/algorithm/string/compare.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <gnutls/crypto.h>
+#include <gnutls/gnutls.h>
+
+#include <memory>
+#include <system_error>
+
+namespace s3 {
+
+constexpr auto algorithm = "AWS4-HMAC-SHA256";
+
+// time_source //
+
+time_source::time_source()
+  : time_source(&default_source, 0) {}
+
+time_source::time_source(timestamp instant)
+  : time_source([instant]() { return instant; }, 0) {}
+
+ss::sstring time_source::format(const char* fmt) const {
+    std::array<char, formatted_datetime_len> out_str{};
+    auto point = _gettime_fn();
+    std::time_t time = std::chrono::system_clock::to_time_t(point);
+    std::tm* gm = std::gmtime(&time);
+    auto ret = std::strftime(out_str.data(), out_str.size(), fmt, gm);
+    vassert(ret > 0, "Invalid date format string");
+    return ss::sstring(out_str.data(), ret);
+}
+
+ss::sstring time_source::format_date() const { return format("%Y%m%d"); }
+
+ss::sstring time_source::format_datetime() const {
+    return format("%Y%m%dT%H%M%SZ");
+}
+
+timestamp time_source::default_source() {
+    return std::chrono::system_clock::now();
+}
+
+// signature_v4 //
+
+enum {
+    sha256_digest_length = 32,
+};
+
+using hmac_digest = std::array<char, sha256_digest_length>;
+
+template<class Container>
+static ss::sstring hexdigest(Container const& digest) {
+    std::array<uint8_t, sha256_digest_length> result{};
+    std::memcpy(result.data(), digest.data(), digest.size());
+    return to_hex(bytes_view{result.data(), result.size()});
+}
+
+static hmac_digest hmac(std::string_view key, std::string_view value) {
+    hmac_digest digest{};
+    int ec = gnutls_hmac_fast(
+      GNUTLS_MAC_SHA256,
+      key.data(),
+      key.size(),
+      value.data(),
+      value.size(),
+      digest.data());
+    vassert(ec >= 0, "HMAC error '{}'", ec);
+    return digest;
+}
+
+static result<ss::sstring> sha_256(std::string_view str) {
```

Comment:
It looks like you could use `hash_sha256` from `src/v/hashing/secure.h`

Reply:
yep